### PR TITLE
chore: pass repo review

### PR DESCRIPTION
**Pull request by Saransh-cpp**
- Modifies `setup.cfg`, `.pre-commit-config.yaml`, and `pyproject.toml` to pass the repo review
- Adds `noxfile.py` for beginners
- Other changes are by the newly added pre-commit hooks

Repo review for this branch - https://scikit-hep.org/repo-review/?repo=Saransh-cpp%2Fvector&branch=saransh%2Frepo-review
(The red cross should ideally not appear since the dependabot check was fixed in the repo review repository. All the other green ticks are very satisfying to look at!)

### CODE REVIEW

File: noxfile.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+"""
```
Comment (Saransh-cpp) [inline noxfile.py:1]: The `noxfile`.

File: pyproject.toml [hunk 2]
```diff
@@ -9,9 +9,13 @@ build-backend = "setuptools.build_meta"
 write_to = "src/vector/version.py"
```
Comment (Saransh-cpp) [inline pyproject.toml:9]: `pyproject.toml`

File: src/vector/_methods.py [hunk 3]
```diff
@@ -2742,7 +2670,7 @@ def _lib_of(*objects: VectorProtocol) -> Module:  # NumPy-like module
         if isinstance(obj, Vector):
             if lib is None:
                 lib = obj.lib
-            elif lib is not obj.lib:
+            elif lib is not obj.lib:  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/_methods.py:2673]: `type: ignore[unreachable]`

File: src/vector/_methods.py [hunk 4]
```diff
@@ -2800,7 +2728,7 @@ def _handler_of(*objects: VectorProtocol) -> VectorProtocol:
             continue
         if handler is None:
             handler = obj
-        elif _get_handler_index(obj) > _get_handler_index(handler):
+        elif _get_handler_index(obj) > _get_handler_index(handler):  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/_methods.py:2731]: `type: ignore[unreachable]`

File: src/vector/_methods.py [hunk 5]
```diff
@@ -2823,7 +2751,7 @@ def _flavor_of(*objects: VectorProtocol) -> typing.Type[VectorProtocol]:
                 is_momentum = False
             if handler is None:
                 handler = obj
-            elif isinstance(obj, VectorObject):
+            elif isinstance(obj, VectorObject):  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/_methods.py:2754]: `type: ignore[unreachable]`

File: src/vector/backends/awkward.py [hunk 6]
```diff
@@ -573,12 +574,12 @@ def elements(self) -> typing.Tuple[ArrayOrRecord]:
...
-        elif issubclass(cls, Vector3D):
-            return "Momentum3D"
-        elif issubclass(cls, Vector4D):
-            return "Momentum4D"
+        if issubclass(cls, Vector2D):  # type: ignore[unreachable]
+            return "Momentum2D"  # type: ignore[unreachable]
+        elif issubclass(cls, Vector3D):  # type: ignore[unreachable]
+            return "Momentum3D"  # type: ignore[unreachable]
+        elif issubclass(cls, Vector4D):  # type: ignore[unreachable]
+            return "Momentum4D"  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/backends/awkward.py:582]: `type: ignore[unreachable]`

File: src/vector/backends/numpy.py [hunk 7]
```diff
@@ -251,7 +251,7 @@ def _toarrays(
     """
     istuple = True
     if not isinstance(result, tuple):
-        istuple = False
+        istuple = False  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/backends/numpy.py:254]: `type: ignore[unreachable]`

File: src/vector/backends/numpy.py [hunk 8]
```diff
@@ -285,14 +285,14 @@ def _shape_of(result: typing.Tuple[FloatArray, ...]) -> typing.Tuple[int, ...]:
         (2,)
     """
     if not isinstance(result, tuple):
-        result = (result,)
+        result = (result,)  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/backends/numpy.py:288]: `type: ignore[unreachable]`

File: src/vector/backends/numpy.py [hunk 9]
```diff
@@ -285,14 +285,14 @@ def _shape_of(result: typing.Tuple[FloatArray, ...]) -> typing.Tuple[int, ...]:
...
     shape = None
     for x in result:
         if hasattr(x, "shape"):
             thisshape = list(x.shape)
         elif isinstance(x, collections.abc.Sized):
             thisshape = [len(x)]
-        if shape is None or thisshape[0] > shape[0]:
+        if shape is None or thisshape[0] > shape[0]:  # type: ignore[unreachable]
```
Comment (Saransh-cpp) [inline src/vector/backends/numpy.py:295]: `type: ignore[unreachable]`

**Review by Saransh-cpp — COMMENTED** [review]
This PR grew bigger than I thought. Most of the changes are due to the updated `pre-commit` configuration, but there are a few changes which I think should be reviewed once. I have marked them down below. @henryiii @jpivarski could you please review them once? Thanks!

File: MANIFEST.in [hunk 10]
```diff
@@ -1,3 +1,3 @@
-include LICENSE README.md
```
Comment (Saransh-cpp) [inline MANIFEST.in:1]: `MANIFEST.in`

File: src/vector/backends/numpy.py (hunk 7 above)
Comment (henryiii) [reply src/vector/backends/numpy.py:254]: Yeah, we just said it was a Tuple, now we are checking to see if it is. Not sure how "publicly" we want to let users know it accepts non Tuples; it technically should be `Union[Tuple[ScalarCollection, ...], ScalarCollection]``.

File: setup.cfg [hunk 11]
```diff
@@ -57,7 +57,7 @@ numba_extensions =
     init = vector:register_numba
 
 [flake8]
-ignore = E203, E231, E501, E722, W503, B950
-select = C,E,F,W,T,B,B9
+extend-ignore = E203, E231, E501, E722, W503, B950, D103, D102, D101, D100, D107, D105, D205, D400, D401, D104, D412
+extend-select = C,E,F,W,T,B,B9
```
Comment (henryiii) [inline setup.cfg:61]:
```suggestion
extend-select = B9
```

The extend doesn't need the "default" values. 85% sure that includes T and B.

File: setup.cfg [hunk 12]
```diff
@@ -57,7 +57,7 @@ numba_extensions =
     init = vector:register_numba
 
 [flake8]
-ignore = E203, E231, E501, E722, W503, B950
-select = C,E,F,W,T,B,B9
+extend-ignore = E203, E231, E501, E722, W503, B950, D103, D102, D101, D100, D107, D105, D205, D400, D401, D104, D412
```
Comment (henryiii) [inline setup.cfg:60]: Some of the basic ones here can be removed. I've forgotten which ones - I'd try removing all the "old" ones and seeing which ones show up.

File: noxfile.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+
+Execute ``nox`` to run the default sessions ("lint", "tests", "doctests")
+and ``nox --session <session>`` to run a specific session.
+"""
+import shutil
+from pathlib import Path
+
+import nox
+
+ALL_PYTHONS = ["3.6", "3.7", "3.8", "3.9"]
```
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply noxfile.py:12]: Missed that. Thanks!

File: setup.cfg (hunk 12 above)
Comment (Saransh-cpp) [reply setup.cfg:60]: Done! Thanks!

File: noxfile.py [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+def lint(session):
+    """Run the linter."""
+    session.install("pre-commit")
+    session.run("pre-commit", "run", "--all-files", *session.posargs)
+
+
+@nox.session(python=ALL_PYTHONS, reuse_venv=True)
+def tests(session):
+    """Run the unit and regular tests."""
+    session.install("-e", ".[awkward,test,test-extras]")
```
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply noxfile.py:30]: Yes, tested it out, working fine!

File: MANIFEST.in (hunk 10 above)
Comment (Saransh-cpp) [reply MANIFEST.in:1]: The SDist workflow was erroring out due to this -
```
check-manifest...........................................................Failed
- hook id: check-manifest
- exit code: 1
lists of files in version control and sdist do not match!
missing from sdist:
  noxfile.py
suggested MANIFEST.in rules:
  include *.py
```
https://github.com/scikit-hep/vector/runs/7136492072?check_suite_focus=true

Is there a way to ignore `noxfile`?

File: noxfile.py [hunk 15]
```diff
@@ -0,0 +1,66 @@
+"""
+Use ``nox`` to run various test suites of vector locally.
```
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply noxfile.py:2]: Oh, I didn't know that. This seems redundant now; I think it should just be removed.

File: noxfile.py [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+            )
+            session.run("python", "-m", "http.server", "8001", "-d", "_build/html")
+        else:
+            print("Unsupported argument to docs")  # noqa: T201
+
+
+@nox.session
+def build(session):
+    """Build an SDist and wheel."""
+    build_p = DIR.joinpath("build")
```
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply noxfile.py:61]: Yes, I will take that up! Would you like it to go in this PR, or should I remove the build session from here, and have the hatchling part go in the next PR?

[the remark this answers is not in the document]
Comment (ofek) [reply noxfile.py:53]: > moving to a modern build system like hatchling [...] I could run ini2toml and get 80% there

Please try out https://hatch.pypa.io/latest/intro/#existing-project

File: MANIFEST.in (hunk 10 above)
Comment (Saransh-cpp) [reply MANIFEST.in:1]: Thank you!

File: noxfile.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (henryiii) [reply noxfile.py:53]: Thanks, I just discovered that recently actually, see https://github.com/scikit-build/moderncmakedomain/pull/5 :). FYI, @Saransh-cpp, all I had to do was run `pipx run hatch new --init` and then delete the `setup.*` files.

We should do things like that in another PR. Best to keep smaller, focused PRs or nothing will be reviewable / mergable. (Thinking about PRs that have taken 2+ years to merge here...) :)

**Review by henryiii — APPROVED** [review]

File: noxfile.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Saransh-cpp) [reply noxfile.py:53]: Thank you for the suggestions, @henryiii, and @ofek! Yes, agreed. This PR already grew uncontrollably due to the pre-commit hooks.

I will move vector to hatchling in the forthcoming PR!

**Pull request merged**